concrete/src/booleans/types/base.rs

```diff
@@ -0,0 +1,193 @@
+use std::borrow::Borrow;
+use std::ops::{BitAnd, BitOr, BitXor};
+
+use concrete_boolean::ciphertext::Ciphertext;
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+use crate::booleans::client_key::GenericBoolClientKey;
+use crate::booleans::parameters::BooleanParameterSet;
+use crate::booleans::server_key::GenericBoolServerKey;
+use crate::global_state::WithGlobalKey;
+use crate::keys::{ClientKey, RefKeyFromKeyChain};
+use crate::traits::{FheDecrypt, FheEncrypt, FheEq};
+
+/// The FHE boolean data type.
+///
+/// To be able to use this type, the cargo feature `booleans` must be enabled,
+/// and your config should also enable the type with either default parameters or custom ones.
+///
+/// # Example
+/// ```rust
+/// use concrete::prelude::*;
+/// #[cfg(feature = "booleans")]
+/// # {
+/// use concrete::{generate_keys, set_server_key, ConfigBuilder, FheBool};
+///
+/// // Enable booleans in the config
+/// let config = ConfigBuilder::all_disabled().enable_default_bool().build();
+///
+/// // With the booleans enabled in the config, the needed keys and details
+/// // can be taken care of.
+/// let (client_key, server_key) = generate_keys(config);
+///
+/// let ttrue = FheBool::encrypt(true, &client_key);
+/// let ffalse = FheBool::encrypt(false, &client_key);
+///
+/// // Do not forget to set the server key before doing any computation
+/// set_server_key(server_key);
+///
+/// let fhe_result = ttrue & ffalse;
+///
+/// let clear_result = fhe_result.decrypt(&client_key);
+/// assert_eq!(clear_result, false);
+/// # }
+/// ```
+#[cfg_attr(doc, cfg(feature = "booleans"))]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[derive(Clone)]
+pub struct GenericBool<P>
+where
+    P: BooleanParameterSet,
+{
+    pub(in crate::booleans) ciphertext: Ciphertext,
+    pub(in crate::booleans) id: P::Id,
+}
+
+impl<P> GenericBool<P>
+where
+    P: BooleanParameterSet,
+{
+    pub(in crate::booleans) fn new(ciphertext: Ciphertext, id: P::Id) -> Self {
+        Self { ciphertext, id }
+    }
+}
+
+impl<P> GenericBool<P>
+where
+    P: BooleanParameterSet,
+    P::Id: WithGlobalKey<Key = GenericBoolServerKey<P>>,
+{
+    pub fn nand(&self, rhs: &Self) -> Self {
+        self.id.with_unwrapped_global_mut(|key| key.nand(self, rhs))
+    }
+
+    pub fn neq(&self, other: &Self) -> Self {
+        self.id.with_unwrapped_global_mut(|key| {
+            let eq = key.xnor(self, other);
+            key.not(&eq)
+        })
+    }
+}
+
+impl<P, B> FheEq<B> for GenericBool<P>
+where
+    B: Borrow<Self>,
```

Comment:
I used `Borrow` so that rhs could be either a `GenericBool<P>`  or a   `&GenericBool<P>`  Borrow has the blanket impl that allows that, but it seems that `Deref<Targer=T> for T` is not implemented.

```rust
use core::borrow::Borrow;
use core::ops::Deref;

struct FheNumber(u32);


fn fake_with_deref<B: Deref<Target=FheNumber>>(l: B) {
    println!("{}", l.0);
}

fn fake_with_borrow<B: Borrow<FheNumber>>(l: B) {
    println!("{}", l.borrow().0);
}

fn main() {

    let l = FheNumber(32);
    fake_with_deref(&l);
    // no Deref<Target=T> for T
    //fake_with_deref(l);
    
    fake_with_borrow(&l);
    fake_with_borrow(l);
}
```